File: storage-aws/src/test/scala/net/kemitix/thorp/storage/aws/UploaderTest.scala
package net.kemitix.thorp.storage.aws

import org.scalamock.scalatest.MockFactory
import org.scalatest.FreeSpec

class UploaderTest extends FreeSpec with MockFactory {

//  val uiChannel: UChannel[Any, UIEvent] = zioMessage => ()
//
//  "upload" - {
//    val aSource: File = Resource(this, "").toFile
//    val aFile: File   = Resource(this, "small-file").toFile
//    val aHash         = MD5Hash.create("aHash")
//    val hashes        = Hashes.create(MD5, aHash)
//    val remoteKey     = RemoteKey.create("aRemoteKey")
//    val localFile =
//      LocalFile.create(aFile, aSource, hashes, remoteKey, aFile.length)
//    val bucket       = Bucket.named("aBucket")
//    val uploadResult = new UploadResult
//    uploadResult.setKey(remoteKey.key)
//    uploadResult.setETag(aHash.hash())
//    val listenerSettings =
//      UploadEventListener.Settings(uiChannel, localFile, 0, 0, batchMode = true)
//    "when no error" in {
//      val expected =
//        Right(StorageEvent.uploadEvent(remoteKey, aHash))
//      val inProgress = new AmazonUpload.InProgress {
//        override def waitForUploadResult: Task[UploadResult] =
//          Task(uploadResult)
//      }
//      new AmazonS3ClientTestFixture {
//        (() => fixture.amazonS3TransferManager.uploader)
//          .when()
//          .returns(_ => UIO.succeed(inProgress))
//        private val result =
//          invoke(fixture.amazonS3TransferManager)(
//            localFile,
//            bucket,
//            listenerSettings
//          )
//        assertResult(expected)(result)
//      }
//    }
//    "when Amazon Service Exception" in {
//      val exception = new AmazonS3Exception("message")
//      val expected =
//        Right(
//          StorageEvent.errorEvent(ActionSummary.upload(remoteKey.key),
//                                  remoteKey,
//                                  exception))
//      val inProgress = new AmazonUpload.InProgress {
//        override def waitForUploadResult: Task[UploadResult] =
//          Task.fail(exception)
//      }
//      new AmazonS3ClientTestFixture {
//        (() => fixture.amazonS3TransferManager.upload)
//          .when()
//          .returns(_ => UIO.succeed(inProgress))
//        private val result =
//          invoke(fixture.amazonS3TransferManager)(
//            localFile,
//            bucket,
//            listenerSettings
//          )
//        assertResult(expected)(result)
//      }
//    }
//    "when Amazon SDK Client Exception" in {
//      val exception = new SdkClientException("message")
//      val expected =
//        Right(
//          StorageEvent.errorEvent(ActionSummary.upload(remoteKey.key),
//                                  remoteKey,
//                                  exception))
//      val inProgress = new AmazonUpload.InProgress {
//        override def waitForUploadResult: Task[UploadResult] =
//          Task.fail(exception)
//      }
//      new AmazonS3ClientTestFixture {
//        (() => fixture.amazonS3TransferManager.upload)
//          .when()
//          .returns(_ => UIO.succeed(inProgress))
//        private val result =
//          invoke(fixture.amazonS3TransferManager)(
//            localFile,
//            bucket,
//            listenerSettings
//          )
//        assertResult(expected)(result)
//      }
//    }
//    def invoke(transferManager: AmazonTransferManager)(
//        localFile: LocalFile,
//        bucket: Bucket,
//        listenerSettings: UploadEventListener.Settings
//    ) = {
//      val program = Uploader
//        .upload(transferManager)(
//          Uploader.Request(localFile, bucket, listenerSettings))
//      val runtime = new DefaultRuntime {}
//      runtime
//        .unsafeRunSync(
//          program
//            .provide(Config.Live))
//        .toEither
//    }
//  }

}

File: storage-aws/src/test/java/net/kemitix/thorp/storage/aws/UploaderTest.java
package net.kemitix.thorp.storage.aws;

import com.amazonaws.SdkClientException;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.transfer.TransferManager;
import com.amazonaws.services.s3.transfer.Upload;
import com.amazonaws.services.s3.transfer.model.UploadResult;
import net.kemitix.thorp.domain.HashType;
import net.kemitix.thorp.domain.*;
import net.kemitix.thorp.filesystem.Resource;
import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.io.File;

import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.doThrow;

@ExtendWith(MockitoExtension.class)
public class UploaderTest
        implements WithAssertions {

    private final File aSource = Resource.select(this, "").toFile();
    private final File aFile = Resource.select(this, "small-file").toFile();
    private final MD5Hash aHash = MD5Hash.create("aHash");
    private final Hashes hashes = Hashes.create(HashType.MD5, aHash);
    private final RemoteKey remoteKey = RemoteKey.create("aRemoteKey");
    private final LocalFile localFile =
            LocalFile.create(aFile, aSource, hashes, remoteKey, aFile.length());
    private final Bucket bucket = Bucket.named("aBucket");
    private final PutObjectRequest request =
            S3Uploader.request(localFile, bucket);

    private final TransferManager transferManager;

    private final S3TransferManager s3TransferManager;

    private final UploadResult uploadResult;
    private final Upload upload;

    public UploaderTest(@Mock TransferManager transferManager,
                        @Mock Upload upload) {
        this.transferManager = transferManager;
        this.upload = upload;
        s3TransferManager = S3TransferManager.create(transferManager);
        uploadResult = new UploadResult();
        uploadResult.setKey(remoteKey.key());
        uploadResult.setETag(aHash.hash());
    }

    @Test
    @DisplayName("when upload then return upload event")
    public void whenUpload_thenUploadEvent() throws InterruptedException {
        //given
        given(transferManager.upload(request))
                .willReturn(upload);
        given(upload.waitForUploadResult()).willReturn(uploadResult);
        //when
        StorageEvent result = S3Uploader.uploader(s3TransferManager).apply(request);
        //then
        assertThat(result).isInstanceOf(StorageEvent.UploadEvent.class);
    }

    @Test
    @DisplayName("when error uploading then return an error storage event")
    public void whenUploadErrors_thenErrorEvent() {
        //given
        doThrow(SdkClientException.class)
                .when(transferManager)
                .upload(request);
        //when
        StorageEvent result = S3Uploader.uploader(s3TransferManager).apply(request);
        //then
        assertThat(result).isInstanceOf(StorageEvent.ErrorEvent.class);
    }

}
